neqo-transport/src/cc/cubic.rs

```diff
@@ -0,0 +1,182 @@
+            ca_epoch_start: None,
+            last_phase_was_tcp: false,
+            tcp_acked_bytes: 0.0,
+        }
+    }
+}
+
+impl Display for Cubic {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Cubic [last_max_cwnd: {}, k: {}, w_max: {}, ca_epoch_start: {:?}]",
+            self.last_max_cwnd, self.k, self.w_max, self.ca_epoch_start
+        )?;
+        Ok(())
+    }
+}
+
+#[allow(clippy::doc_markdown)]
+impl Cubic {
+    /// Original equations is:
+    /// K = cubic_root(W_max*(1-beta_cubic)/C) (Eq. 2 RFC8312)
+    /// W_max is number of segments of the maximum segment size (MSS).
+    ///
+    /// K is actually the time that W_cubic(t) = C*(t-K)^3 + W_max (Eq. 1) would
+    /// take to increase to W_max. We use bytes not MSS units, therefore this
+    /// equation will be: W_cubic(t) = C*MSS*(t-K)^3 + W_max.
+    ///
+    /// From that equation we can calculate K as:
+    /// K = cubic_root((W_max - W_cubic) / C / MSS);
+    fn calc_k(&self, curr_cwnd: f64) -> f64 {
+        ((self.w_max - curr_cwnd) / CUBIC_C / MAX_DATAGRAM_SIZE_F64).cbrt()
+    }
+
+    /// W_cubic(t) = C*(t-K)^3 + W_max (Eq. 1)
+    fn w_cubic(&self, t: f64) -> f64 {
+        CUBIC_C * (t - self.k).powi(3) * MAX_DATAGRAM_SIZE_F64 + self.w_max
+    }
+}
+
+impl WindowAdjustment for Cubic {
+    // This is because of the cast in the last line from f64 to usize.
+    #[allow(clippy::cast_possible_truncation)]
+    #[allow(clippy::cast_sign_loss)]
+    fn on_packets_acked(
+        &mut self,
+        curr_cwnd: usize,
+        acked_bytes: usize,
+        min_rtt: Duration,
+        now: Instant,
+    ) -> usize {
+        let curr_cwnd_f64 = f64::try_from(u32::try_from(curr_cwnd).unwrap()).unwrap();
+        self.tcp_acked_bytes += f64::try_from(u32::try_from(acked_bytes).unwrap()).unwrap();
+        if self.ca_epoch_start.is_none() {
+            // This is a start of a new congestion avoidance phase.
+            self.ca_epoch_start = Some(now);
+            // reset acked_bytes and estimated_tcp_cwnd;
+            self.tcp_acked_bytes = f64::try_from(u32::try_from(acked_bytes).unwrap()).unwrap();
+            self.estimated_tcp_cwnd = curr_cwnd_f64;
+            if self.last_max_cwnd <= curr_cwnd_f64 {
+                self.w_max = curr_cwnd_f64;
+                self.k = 0.0;
+            } else {
+                self.w_max = self.last_max_cwnd;
+                self.k = self.calc_k(curr_cwnd_f64);
+            }
+            qtrace!([self], "New epoch");
+        }
+
+        let time_ca = (now + min_rtt
+            - if let Some(t) = self.ca_epoch_start {
+                t
+            } else {
+                now
+            })
+        .as_secs_f64();
+        let target = self.w_cubic(time_ca);
+
+        let mut cnt = if target > curr_cwnd_f64 {
+            curr_cwnd_f64 / (target - curr_cwnd_f64) * MAX_DATAGRAM_SIZE_F64
+        } else {
+            100.0 * curr_cwnd_f64
```

Comment:
This is from the linux implementation. It is a corner case, I do not know how to trigger it. This should cause a very small increase in cwnd.
I will explain cnt than this will be more understandable:

cnt in NewReno is always cwnd, i.e. a congestion controller need to receive cwnd of acked bytes to increase cwnd by 1 MSS. For cubic it is calculated depending on W_cubic and tcp estimate, but is still "how many acked bytes it needs to receive to increase cnwd by 1 MSS.
This is leaned on the linux implementation. I may change it to be more similar to the RFC, I think that will make increase slightly slower(tiny, tiny slower).